include/psen_scan_v2/diagnostics.h

```diff
@@ -0,0 +1,184 @@
+// Copyright (c) 2020 Pilz GmbH & Co. KG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+#ifndef PSEN_SCAN_V2_DIAGNOSTICS_H
+#define PSEN_SCAN_V2_DIAGNOSTICS_H
+
+#include <array>
+#include <string>
+#include <map>
+#include <vector>
+#include <set>
+
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+
+#include "psen_scan_v2/scanner_ids.h"
+
+namespace psen_scan_v2
+{
+static constexpr uint32_t DIAGNOSTIC_MESSAGE_RAW_LENGTH_FOR_ONE_DEVICE_IN_BYTES{ 9 };
+static constexpr uint32_t DIAGNOSTIC_MESSAGE_UNUSED_OFFSET_IN_BYTES{ 4 };
+static constexpr uint32_t DIAGNOSTIC_DATA_LENGTH_IN_BYTES{ DIAGNOSTIC_MESSAGE_UNUSED_OFFSET_IN_BYTES +
+                                                           DIAGNOSTIC_MESSAGE_RAW_LENGTH_FOR_ONE_DEVICE_IN_BYTES *
+                                                               sizeof(VALID_SCANNER_IDS) };
+
+enum class DiagnosticCode
+{
+  OSSD1_OC,
+  OSSD_SHRT_C,
+  OSSD_INTEGR,
+  INT,
+  WIN_CLN_AL,
+  POWER_SUPPLY,
+  NETW_PRB,
+  DUST_CRC_FL,
+  OSSD2_OVERCUR,
+  MEAS_PROB,
+  INCOHERENCE,
+  ZONE_INVAL_TRANS,
+  ZONE_INVALID_CONF,
+  WIN_CLN_WARN,
+  INT_COM_PRB,
+  GENERIC_ERR,
+  DISP_COM_PRB,
+  TEMP_MEAS_PROB,
+  ENCOD_OOR,
+  EDM2_ERR,
+  EDM1_ERR,
+  CONF_ERR,
+  OUT_OF_RANGE_ERR,
+  TEMP_RANGE_ERR,
+  ENCODER_GENERIC_ERR,
+  _,
+  UNUSED
+};
+
+class ErrorLocation
+{
+public:
+  using Byte_Location = size_t;
+  using Bit_Location = size_t;
+  ErrorLocation(Byte_Location byte, Bit_Location bit) : byte_(byte), bit_(bit){};
+  inline Byte_Location getByte() const
+  {
+    return byte_;
+  };
+  inline Bit_Location getBit() const
+  {
+    return bit_;
+  };
+
+private:
+  Byte_Location byte_;
+  Bit_Location bit_;
+};
+
+class MonitoringFrameDiagnosticMessage
+{
+public:
+  MonitoringFrameDiagnosticMessage(ScannerId id, ErrorLocation location_);
+
+  bool operator==(const MonitoringFrameDiagnosticMessage& rhs) const;
+
+  friend std::array<uint8_t, DIAGNOSTIC_DATA_LENGTH_IN_BYTES>
+  serializeDiagnosticMessages(std::vector<MonitoringFrameDiagnosticMessage>& messages);
+
+  ScannerId getScannerId() const
+  {
+    return id_;
+  }
+
+  DiagnosticCode getDiagnosticCode() const
+  {
+    return code_;
+  }
+
+  ErrorLocation getErrorLocation() const
+  {
+    return error_location_;
+  }
+
+private:
+  ScannerId id_;
+  DiagnosticCode code_;
```

Comment:
Why is this a member? It can be easily determined when requested, via:
```
error_bits.at(byte_location).at(bit_location);
```
This violates the DRY principle.

Reply:
True. Did not see this.

Reply:
On it